gglasso/solver/single_admm_solver.py

```diff
@@ -7,154 +7,212 @@
     """
     This is an ADMM algorithm for solving the Single Graphical Lasso problem
-    
+
     Omega_0 : start point -- must be specified as a (p,p) array
     S : empirical covariance matrix -- must be specified as a (p,p) array
-    
+
     latent: boolean to indidate whether low rank term should be estimated
     mu1: low rank penalty paramater, if latent=True
-    
-    
+
+
     In the code, X are the SCALED (with 1/rho) dual variables, for the KKT stop criterion they have to be unscaled (i.e. take rho*X) again!
     """
     assert Omega_0.shape == S.shape
     assert S.shape[0] == S.shape[1]
     assert lambda1 > 0
-        
+
     if latent:
         assert mu1 is not None
         assert mu1 > 0
-        
-    (p,p) = S.shape
-    
+
+    (p, p) = S.shape
+
     assert rho > 0, "ADMM penalization parameter must be positive."
-        
-    
-    # initialize 
+
+    # initialize
     status = 'not optimal'
     Omega_t = Omega_0.copy()
     if len(Theta_0) == 0:
         Theta_0 = Omega_0.copy()
     if len(X_0) == 0:
-        X_0 = np.zeros((p,p))
+        X_0 = np.zeros((p, p))
 
     Theta_t = Theta_0.copy()
-    L_t = np.zeros((p,p))
+    Theta_t_1 = Theta_0.copy()
+    L_t = np.zeros((p, p))
     X_t = X_0.copy()
-       
+
     runtime = np.zeros(max_iter)
     kkt_residual = np.zeros(max_iter)
-    
+    first_pass = True
+
     for iter_t in np.arange(max_iter):
-        
-        eta_A = ADMM_stopping_criterion(Omega_t, Theta_t, L_t, rho*X_t, S , lambda1, latent, mu1)
-        kkt_residual[iter_t] = eta_A
-        
         if measure:
             start = time.time()
-                
-        if eta_A <= eps_admm:
+
+        if not first_pass:
+            eta_A = ADMM_stopping_criterion(Omega_t, Theta_t, Theta_t_1, L_t, rho * X_t, S, lambda1, latent, mu1, eps_rel, eps_abs)
+            kkt_residual[iter_t] = eta_A['value']
+        else:
+            eta_A = Fabian_stopping_criterion(Omega_t, Theta_t, L_t, rho * X_t, S, lambda1, latent, mu1)
+            kkt_residual[iter_t] = eta_A['value']
+
+        if len(eta_A["status"]) > 1:  # both primal and dual solutions are optimal
             status = 'optimal'
             break
         if verbose:
             print(f"------------Iteration {iter_t} of the ADMM Algorithm----------------")
-        
+
         # Omega Update
-        W_t = Theta_t - L_t - X_t - (1/rho) * S
+        W_t = Theta_t - L_t - X_t - (1 / rho) * S
         eigD, eigQ = np.linalg.eigh(W_t)
-        Omega_t= phiplus(W_t, beta = 1/rho, D = eigD, Q = eigQ)
-        
+        Omega_t = phiplus(W_t, beta=1 / rho, D=eigD, Q=eigQ)
+
         # Theta Update
-        Theta_t = prox_od_1norm(Omega_t + L_t + X_t, (1/rho)*lambda1)
-        
+        Theta_t = prox_od_1norm(Omega_t + L_t + X_t, (1 / rho) * lambda1)
+        Theta_t_1 = Theta_t.copy()
```

Comment:
If you update `Theta_t_1` after `Theta_t`, they will always be equal. This is also the reason for your issue. It's a bit tricky in the first iteration though, we may need to move the stop_criterion to the end of the iteration (or skip stop_criterion at first iter)